### PR TITLE
Implement intersection of functional volume - fix #26

### LINKED ISSUE #26: New index: intersection between convex hulls
As ecologists may be interested in the intersection of different convex hulls fundiversity could provide a wrapper around `geometry::intersectn()` that outputs the volume of the intersection between two volumes.

The input of this wrapper would be similar to `fd_fric()` (site-species matrix, trait matrix) and the output would be a distance matrix or rather a tidy data.frame with the first two columns giving the ids of the two considered sites, the third column would give the volume of the intersection.

Computationnally this could be quite intensive but should parallelize without issue (see #13).

### PR DESCRIPTION
**Pull request by Rekyt**
fix #26

### CODE REVIEW

**Comment by Rekyt** [comment]
For this PR I still need to:
1. add documentation in the introduction vignette,
2. add performance estimation in the performance vignette,
3. add a memoised version of fd_chull_intersect().

**Comment by Rekyt** [comment]
@Bisaloo the PR is mostly ready for review.
One point that we have to inspect more closely is the risk of errors from `geometry::instersect()` it seems that many edge cases may fail and the documentation of this function clearly says that it sometimes have bad ouput.
So we should test the function on as many cases as possible and probably warn the users in the documentation.

File: R/fd_chull_intersect.R [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  traits2 <- traits2[!duplicated(traits2),, drop = FALSE]
+
+  if (ncol(traits1) == 1L & ncol(traits2) == 1L) {
+
+    r1 <- range(traits1)
+    r2 <- range(traits2)
+
+    # Range of the overlap
+    r3 <- c(max(r1[1], r2[1]),
+            min(r1[2], r2[2]))
```
Comment (Bisaloo) [inline R/fd_chull_intersect.R:15]:
```suggestion
    # Range of the overlap
    r3 <- c(max(traits1, traits2),
            min(traits1, traits2))
```

File: R/fd_chull_intersect.R [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+    r_overlap <- diff(r3)
```
Comment (Bisaloo) [inline R/fd_chull_intersect.R:17]:
```suggestion
    r_overlap <- r3[2] - r3[1]
```

Better performance:

|expr               |  min|   lq|      mean| median|   uq|     max| neval|
|:------------------|----:|----:|---------:|------:|----:|-------:|-----:|
|{     1 - 2 }      |   84|  127|  161.1078|    137|  155| 4997253| 1e+06|
|{     diff(1, 2) } | 2214| 2407| 2760.8737|   2509| 2753| 9890056| 1e+06|

File: R/fd_chull_intersect.R [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+    r_overlap <- ifelse(r_overlap < 0, NA_real_, r_overlap)
```
Comment (Bisaloo) [inline R/fd_chull_intersect.R:19]: I don't understand how this case could happen. Could you add a comment why this check is necessary please?

File: R/fd_chull_intersect.R [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        ch = list(
+          "hull" = seq_len(nrow(traits1) + nrow(traits2)),
+          "area" = NA_real_,
+          "vol" = NA_real_
+        )
+      )
+    )
+  }
+
+  return(geometry::intersectn(traits1, traits2, options = "FA"))
```
Comment (Bisaloo) [inline R/fd_chull_intersect.R:44]: What about return only the `ch` element here? We would not have to create a "complex" nested list for the edge cases then

File: R/fd_fric_intersect.R [hunk 5]
```diff
@@ -0,0 +1,90 @@
+#' Intersection between functional volume of pairs of sites
```
Comment (Bisaloo) [inline R/fd_fric_intersect.R:1]: The term functional volume is not super clear / intuitive for me. Is it a term that is present in the literature and easy to understand for people in the field? If so, we can leave it as is. Otherwise, it might be easier to understand if we re-use the term functional richness

File: R/fd_fric_intersect.R [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+#' thus standardized self-intersection volumes can be lower than one.
+#'
+#' @inheritParams fd_fric
+#'
+#' @return a data.frame with three columns:
+#' * `first_site` the names of the first site used in the pair `sp_com`,
+#' * `second_site` the names of the first site used in the pair,
+#' * `FRic_intersect` the values of intersection of functional volumes of each
+#' pair of site.
+#'
```
Comment (Bisaloo) [inline R/fd_fric_intersect.R:16]:
```suggestion
#'
#' @seealso [fd_fric()], [geometry::intersectn()]
#'
```

File: R/fd_fric_intersect.R [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+  }
+
+  max_range <- 1
+
+  if (stand) {
+    max_range <- fd_chull(traits)$vol
+  }
+
+  # All pairs of sites (not within themselves)
+  all_site_comb <- tryCatch({
```
Comment (Bisaloo) [inline R/fd_fric_intersect.R:62]: Can you add a comment explaining why the `tryCatch()` is necessary please?

File: R/fd_fric_intersect.R [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    t(combn(rownames(sp_com), 2))
+  },
+  error = function(x, msg) {
+    return(numeric(0))
+  })
+
+  self_intersection <- matrix(rep(rownames(sp_com), each = 2),
```
Comment (Bisaloo) [inline R/fd_fric_intersect.R:69]: We know that the self intersection will be the itself so we should compute it why `convhulln()` instead of `intersectn()` because it's much more efficient:

|expr                 |      min|        lq|      mean|   median|        uq|      max| neval|
|:--------------------|--------:|---------:|---------:|--------:|---------:|--------:|-----:|
|intersectn(ps1, ps1) | 1312.947| 1370.2895| 1457.8260| 1395.387| 1431.6545| 6963.097|   100|
|convhulln(ps1)       |  183.773|  197.8175|  211.3715|  209.076|  224.3845|  272.437|   100|

**Review by Bisaloo — CHANGES_REQUESTED** [review]

File: R/fd_fric_intersect.R (hunk 5 above)
Comment (Rekyt) [reply R/fd_fric_intersect.R:1]: If it's unclear to you than we should change it. Because even though it is straightfoward in functional ecology, if we want the package to target a broader group than we should use more commonly used terms. What alternative do you propose? Because "functional richness intersection" is not that common and does not it seem clearer to me. "Intersection of convex hulls"?

File: R/fd_fric_intersect.R (hunk 7 above)
Comment (Rekyt) [reply R/fd_fric_intersect.R:62]: the tryCatch was sort of a temporary. We need a little more testing here because the tryCatch() has to handle errors from `geometry::intersectn()` when it cannot find points in which the intersection is used. Have a look at `geometry::feasible.points()`.

File: R/fd_fric_intersect.R (hunk 8 above)
Comment (Rekyt) [reply R/fd_fric_intersect.R:69]: Indeed!

File: R/fd_fric_intersect.R (hunk 5 above)
Comment (Bisaloo) [reply R/fd_fric_intersect.R:1]: We can keep functional volume but add in parenthesis convex hull or something like this maybe?

File: R/fd_fric_intersect.R (hunk 7 above)
Comment (Bisaloo) [reply R/fd_fric_intersect.R:64]: But the `tryCatch()` here only wraps `combn()`, no?

File: R/fd_chull_intersect.R (hunk 3 above)
Comment (Rekyt) [reply R/fd_chull_intersect.R]: fixed by fc3de5c

File: R/fd_chull_intersect.R (hunk 4 above)
Comment (Rekyt) [reply R/fd_chull_intersect.R]: fixed by c922b62

File: R/fd_fric_intersect.R (hunk 7 above)
Comment (Rekyt) [reply R/fd_fric_intersect.R:64]: Oops yes indeed! I was thinking about a different piece of code.
In this case it is to manage the case where there is a single row name.
Because `combn("a", 2)` returns an error.

File: R/fd_fric_intersect.R (hunk 8 above)
Comment (Rekyt) [reply R/fd_fric_intersect.R:70]: Fixed by 5581964

File: R/fd_fric_intersect.R (hunk 7 above)
Comment (Rekyt) [reply R/fd_fric_intersect.R]: Fixed by 5581964

File: R/fd_fric_intersect.R (hunk 5 above)
Comment (Rekyt) [reply R/fd_fric_intersect.R]: Fixed by 8cd4668

**Review by Rekyt — COMMENTED** [review]
I normally solved all the issues you pointed at!

**Review by Bisaloo — APPROVED** [review]
